### ISSUE TITLE
Potential forced style layout

### ISSUE THREAD

**Issue by bakura10** [akamai/boomerang#348]
Hi :),

We are creating themes on Shopify platform, which uses internally Boomerang for its performance measuring.

While looking for performance bottlenecks, I found Chrome reporting several "forced recalculation" and "forced layout", that were all attributed to Boomerang and specifically this part: https://github.com/akamai/boomerang/blob/e7f702e5362e8a8f56f0cf3040ee24077d13a9a8/plugins/restiming.js#L833

I am not super familiar with Boomerang code base and how it is supposed to work, but reading the code seems to indicate that, inside a for loop, you are both asking the browser dimensions: https://github.com/akamai/boomerang/blob/e7f702e5362e8a8f56f0cf3040ee24077d13a9a8/plugins/restiming.js#L833

And then invalidate by writing the DOM here: https://github.com/akamai/boomerang/blob/e7f702e5362e8a8f56f0cf3040ee24077d13a9a8/plugins/restiming.js#L861

This would therefore cause as many invalidations as you have items in the for loop. In our case, Boomerang seems to cause a bit of overhead due to this.

The read should probably be batched, and then all the images should be updated after all the reading has happened, to avoid those forced reflows.

But, as said, I am not familiar enough with this codebase, there might be reasons this is done this way.

**Comment by bluesmoon** [comment]
Line 861 doesn't update the DOM. It operates on an in-memory image object that isn't attached to the DOM. It's only used for srcsets.

**Comment by bakura10** [comment]
Thanks for the answer. I will do some more debugging then, because the more `<picture>` tag I add on the page, the more forced reflows attributed to the getBoundingClientRect appears in Chrome inspector tools. May be a side effect of another code. I will let you know if I find anything :)

**Comment by bluesmoon** [comment]
The getBoundingClientRect probably is causing a reflow, but it isn't being invalidated later. Are the picture elements visible when this happens?

**Comment by bakura10** [comment]
@bluesmoon , yes, all the picture element are inside a slideshow. For animation purpose, each slide is using visibility: hidden (so probably considered as "visible" as the space is allocated). Each slide contains a <picture> tag.

Chrome inspector reports me one reflow _per slide_. So with six slides, I get 6 reflows reported on the getBoundingClientRect in Boomerang, which quickly adds up, performance wise.

I will do further debugging on my end.

**Comment by bakura10** [comment]
For information if this may helped for debugging @bluesmoon , here is the store where I am facing this issue: https://prestige-theme-allure.myshopify.com/

You can see in the inspector tool a succession of recalculation and forced layout in quick succession in Boomerang code:

<img width="725" alt="image" src="https://user-images.githubusercontent.com/1198915/234508329-f8c224a2-8493-4e41-a9df-138e68224150.png">

However, I found some code that may be not optimal in Shopify side as well, but I am having a hard time trying to find the root cause of this here.

**Comment by nicjansma** [comment]
Note changing the image source with `realImg.src = src` should only be happening if the page has opted into `ResTiming.getSrcsetDimensions` which is `false` by default.

https://github.com/akamai/boomerang/commit/957f2b799edf905326c737f6afb31de578849c8d

It's possible the Shopify version/fork didn't have this fix.  I think Shopify has moved off Boomerang, but for anyone else seeing issues with forced layouts, please review if the above config is set.

If there are other cases we're causing layout thrashing, please open a fresh ticket.
